### PR TITLE
Update for "multiline processing for tail input plugin"

### PR DESCRIPTION
**Pull request by semigroupoid**
This PR contains the content of the existing PR [5603](https://github.com/influxdata/telegraf/pull/5603) but rebased on top of the current master branch of this repository.

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by zibuyu1995** [comment]
@ssoroka Hello, would really like to see this PR merged and released. Any idea on when that may be ?

File: plugins/inputs/tail/README.md
Comment (ssoroka) [inline plugins/inputs/tail/README.md:63]: This could be clearer. "what" is already not a great name (I understand it was borrowed from elastic; something like "MatchWhichLine" would be better). You could add some context and examples; eg "previous means any line matching the pattern belongs to the previous line"

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:51]: This isn't necessary. Looks like it's here from a bad merge. (FYI You would probably have fewer problems using rebase instead of merge)

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:247]: I don't really like that receiverMultiline forks the behavior here. It definitely adds a lot of code complexity and likely introduces bugs in how the two receivers process input. eg, search for "Block until plugin is stopping or room is available to add metrics" and you'll find that this other receiver code has already changed, and multiline isn't supporting coordination with downstream outputs to not overwhelm them. Now the multiline receiver is out of sync and won't respect this behavior. In the best case scenario, future updates will have to update two versions of the code.

How difficult do you think it'd be to merge the two receiver functions? Any concerns?

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:332]: can define this once and then do `Reset(t.MultilineConfig.Timeout.Duration)`

File: plugins/inputs/tail/README.md
Comment (semigroupoid) [reply plugins/inputs/tail/README.md:63]: I agree on the unclear naming. I renamed the config element and the added more description for what it accomplishes.

**Comment by semigroupoid** [comment]
Hi @ssoroka, thanks for your detailed review. I have merged the two receiver functions into one, respecting the changes regarding downstream consumers.

File: plugins/inputs/tail/tail.go
Comment (semigroupoid) [reply plugins/inputs/tail/tail.go:335]: Not sure if this is equivalent to the original code, but to me this should be the correct behavior: If the channel is closed (i.e. the channel receive returned nil and channel closed) *and* the multiline buffer is empty, there is no more processing to do and we can quit the function.

**Comment by zibuyu1995** [comment]
When you plan to merge this PR?

**Comment by ssoroka** [comment]
Waiting on a review from @danielnelson _*poke*_

**Comment by zibuyu1995** [comment]
@danielnelson  We‘d really appreciate if you could review the code.

**Comment by semigroupoid** [comment]
@ssoroka, @binchow-ai I've fixed the conflicts with the current master branch, this should be ready to be merged now.

**Comment by kemuning** [comment]
Any example of the syntax how to use this multiline feature ?

**Comment by zibuyu1995** [comment]
@kemuning 
This is my use case:
```
[[inputs.tail]]
  files = ["/opt/emqx/log/emqx.log.[1-9]"]
  from_beginning = false
  data_format = "grok"
  grok_patterns = ['^%{TIMESTAMP_ISO8601:timestamp:ts-"2006-01-02 15:04:05.000"} \[%{LOGLEVEL:level}\] (?m)%{GREEDYDATA:messages}$']
  [inputs.tail.multiline]
   pattern = '^%{TIMESTAMP_ISO8601:timestamp:ts-"2006-01-02 15:04:05.000"} \[%{LOGLEVEL:level}\]'
   match_which_line = "previous"
   negate=true
   timeout = "1s"
```

**Comment by kemuning** [comment]
@zibuyu1995 Thanks.

**Comment by kemuning** [comment]
@zibuyu1995 Strange ... I have this config : 
```
[inputs.tail.multiline]
   pattern = '^%{GREEDYDATA:logmessage}'
   match_which_line = "previous"
   negate=true
   timeout = "1s"
```

but then I got an error ```Error parsing tail, line 70: field corresponding to `multiline' is not defined in tail.Tail ```, where line 70 is ```[inputs.tail.multiline]```.

Any idea ?

**Comment by zibuyu1995** [comment]
@kemuning No, version 1.5.2 does not support multiline feature , you need to select https://github.com/semigroupoid/telegraf

**Comment by vtamilselvan141** [comment]
Is there any timeline when will this PR be merged ?

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:298]: We can't remove this, or the plugin can hang and not shutdown properly.

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:330]: Not sure what the compiler does here, but it looks like this is being reallocated every iteration, where we could just be setting it to nil.

File: plugins/inputs/tail/tail.go
Comment (ssoroka) [inline plugins/inputs/tail/tail.go:133]: How do you feel about changing this name to `invert_match`?

**Comment by ssoroka** [comment]
Would like to see a few more things before this gets merged. Also if anyone is really interested in it, please check out this branch and build it, try it out locally and see how it works for you.

**Comment by lsitzmann** [comment]
Hello everybody interested in this feature.

we are using this feature (build locally) in a production environment for more than one year by now. 
We are gathering logfiles from about 20 machines. Currently at a rate of about 250k lines per hour. 
About 1% of the lines are "multilines".
The feature simply works without any problems. 
I really wouldn't want to miss this feature and I'm looking forward to switch to an officially supported release.

So please merge this PR soon.

**Comment by ssoroka** [comment]
@lsitzmann I really appreciate hearing about your experience! Your vote of confidence means a lot.
@semigroupoid I'm happy to make any remaining changes if you'd like me to get it merged in quickly.

**Comment by vtamilselvan141** [comment]
Is there any updates on this PR ?

**Comment by sjwang90** [comment]
Hey @semigroupoid, would you like make the changes @ssoroka mentioned above.  If anyone would like to check out this branch and complete the fixes that would be awesome too.  We'd love to get this merged in before the next release.

**Review by ssoroka — APPROVED** [review]

**Comment by ssoroka** [comment]
Look like this introduces some kind of race in TestGrokParseLogFilesWithMultilineTailerCloseFlushesMultilineBuffer test.
 
```
panic: test timed out after 10m0s

goroutine 65 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1377 +0x11c
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x52

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc0001ac000, 0xc2d855, 0x43, 0xc368c8, 0x1)
	/usr/local/go/src/testing/testing.go:961 +0x68a
testing.runTests.func1(0xc0001ac000)
	/usr/local/go/src/testing/testing.go:1202 +0xa7
testing.tRunner(0xc0001ac000, 0xc00015fd40)
	/usr/local/go/src/testing/testing.go:909 +0x19a
testing.runTests(0xc000106760, 0x1097300, 0x16, 0x16, 0x0)
	/usr/local/go/src/testing/testing.go:1200 +0x522
testing.(*M).Run(0xc000154400, 0x0)
	/usr/local/go/src/testing/testing.go:1117 +0x300
main.main()
	_testmain.go:86 +0x224

goroutine 146 [sync.Cond.Wait, 9 minutes]:
runtime.goparkunlock(...)
	/usr/local/go/src/runtime/proc.go:310
sync.runtime_notifyListWait(0xc0004fa050, 0x3)
	/usr/local/go/src/runtime/sema.go:510 +0xf8
sync.(*Cond).Wait(0xc0004fa040)
	/usr/local/go/src/sync/cond.go:56 +0x8e
github.com/influxdata/telegraf/testutil.(*Accumulator).Wait(0xc000480000, 0x4)
	/go/src/github.com/influxdata/telegraf/testutil/accumulator.go:333 +0xb8
github.com/influxdata/telegraf/plugins/inputs/tail.TestGrokParseLogFilesWithMultilineTailerCloseFlushesMultilineBuffer(0xc000474000)
	/go/src/github.com/influxdata/telegraf/plugins/inputs/tail/tail_test.go:237 +0x4b9
testing.tRunner(0xc000474000, 0xc368c8)
	/usr/local/go/src/testing/testing.go:909 +0x19a
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x652

goroutine 50 [select, 9 minutes]:
github.com/influxdata/tail/watch.(*InotifyTracker).run(0xc000242040)
	/go/pkg/mod/github.com/influxdata/tail@v1.0.1-0.20200707181643-03a791b270e4/watch/inotify_tracker.go:224 +0x2f4
created by github.com/influxdata/tail/watch.glob..func1
	/go/pkg/mod/github.com/influxdata/tail@v1.0.1-0.20200707181643-03a791b270e4/watch/inotify_tracker.go:54 +0x28a

goroutine 31 [syscall, 9 minutes]:
syscall.Syscall6(0xe8, 0x7, 0xc0001eeb64, 0x7, 0xffffffffffffffff, 0x0, 0x0, 0xc000036500, 0xc0001eeae8, 0x46107c)
	/usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
golang.org/x/sys/unix.EpollWait(0x7, 0xc0001eeb64, 0x7, 0x7, 0xffffffffffffffff, 0x8ce01c, 0x468430, 0x411a90)
	/go/pkg/mod/golang.org/x/sys@v0.0.0-20200212091648-12a6c2dcc1e4/unix/zsyscall_linux_amd64.go:1881 +0x86
gopkg.in/fsnotify%2ev1.(*fdPoller).wait(0xc000122740, 0xc000008000, 0x1, 0x10c4520)
	/go/pkg/mod/gopkg.in/fsnotify.v1@v1.4.7/inotify_poller.go:86 +0xb9
gopkg.in/fsnotify%2ev1.(*Watcher).readEvents(0xc000118a00)
	/go/pkg/mod/gopkg.in/fsnotify.v1@v1.4.7/inotify.go:192 +0x2b2
created by gopkg.in/fsnotify%2ev1.NewWatcher
	/go/pkg/mod/gopkg.in/fsnotify.v1@v1.4.7/inotify.go:59 +0x31d
FAIL	github.com/influxdata/telegraf/plugins/inputs/tail	600.021s
```

I'

**Comment by ssoroka** [comment]
Ok, I have a fix for this failure. I've moved it into another PR to keep it clean. https://github.com/influxdata/telegraf/pull/8167

**Comment by ssoroka** [comment]
merged in https://github.com/influxdata/telegraf/pull/8190

**Pull request closed without merging**
